src/ApiKitClient.ts

```diff
@@ -0,0 +1,56 @@
+import axios, { AxiosError, AxiosRequestConfig } from 'axios';
+import { HttpMethod, ApiClientResponse, PaginatedResponse } from './types';
+
+export function ApiKitClient<T>(
+  baseUrl: string,
+  authToken?: string,
+  handleUnauthorizedAccess?: () => void
+) {
+  const axiosInstance = axios.create({
+    baseURL: baseUrl,
+    headers: {
+      ...(authToken ? { 'Authorization': `Bearer ${authToken}` } : {}),
+    },
+  });
+
+  axiosInstance.interceptors.response.use(response => response, (error: AxiosError) => {
+    if (error.response?.status === 401 && handleUnauthorizedAccess) {
+      handleUnauthorizedAccess();
+    }
+    return Promise.reject(error);
+  });
+
+  const request = async <R = T>(
+    method: HttpMethod,
+    endpoint: string,
+    body?: T,
+    params?: URLSearchParams
+  ): Promise<ApiClientResponse<R>> => {
+    const config: AxiosRequestConfig = {
+      url: endpoint,
+      method,
+      data: body,
+      params,
+    };
+
+    try {
+      const response = await axiosInstance.request(config);
+      return { status: response.status, data: response.data };
+    } catch (error) {
+      const axiosError = error as AxiosError | any;
+      return {
+        status: axiosError.response?.status || 500,
+        errors: axiosError.response?.data.errors,
+      };
+    }
+  };
```

Comment:
where is the logic for `handleUnauthorizedAccess`

Reply:
This is the point! (To not be logic here)...
As, we agreed on the call with Nawwar, that the logic of unauthorization should be a callback to allow the consumer app to determine what is the logic should be, for ex:

```js
Router.push('/login') // in React or Next
```

Reply:
aha so the logic is not handled in this app

Reply:
Exactly!